### PR TITLE
ENH: stats.yeojohnson_normmax: add parameter ymax

### PR DESCRIPTION
**Pull request by xuefeng-xu**
#### Reference issue

#19631 and #19016

#### What does this implement/fix?

- Remove the magic number 20
- Add `ymax` similar to `boxcox_normmax`

#### Additional information

- If `ymax=np.inf`, `yeojohnson_normmax` compute the true, unconstrained optimal lambda.
- By default, `ymax` is the maximum float of the input data's dtype to avoid overflow.

### CODE REVIEW

**Comment by mdhaber** [comment]
@xuefeng-xu Does sklearn still need this, or what is the motivation now?

**Comment by xuefeng-xu** [comment]
@mdhaber Thanks for asking. My motivation is to finish #19016 since the issue in `yeojohnson_llf` has been fixed in #24172. Also, this PR does the same thing as in `boxcox_normmax`, it's good to keep them similar. I'm not sure about scikit-learn.

**Comment by mdhaber** [comment]
> Also, this PR does the same thing as in `boxcox_normmax`, it's good to keep them similar

It would be if `boxcox_normmax` set a good precedent. The presence of the `brack` and `optimize` options and `method='all'` have made it much more challenging to add an `axis` argument and array API support (for consistency with all other reducing functions in `scipy.stats`), and I don't agree with the choice of `ymax`'s default behavior now. (Yes...even though I was involved in adding some of these things.)

For these and similar reasons, I have been become uncomfortable with the interface of SciPy's Box-Cox / Yeo-Johnson transforms and related plot tests, and I'd rather put time into rethinking them to address real needs than adding options to the existing functions.

Do you need this option for your work, or do you know anyone who does?

**Comment by xuefeng-xu** [comment]
> I'd rather put time into rethinking them to address real needs than adding options to the existing functions.

I see, thanks for letting me know. I agree with your thought.

> Do you need this option for your work, or do you know anyone who does?

No, I just feel the current `yeojohnson_normmax` doesn't do the right thing.

Based on that, feel free to close this PR. I'd like to hear more from you and open a PR in the future.

**Comment by mdhaber** [comment]
> No, I just feel the current `yeojohnson_normmax` doesn't do the right thing.

Would the right thing be to return the constrained or unconstrained MLE? (Or you mean you want the option to do either?)

**Comment by xuefeng-xu** [comment]
My intuition is to return the unconstrained MLE since there's no ambiguity. Maybe other library (sklearn) can do it with constrained one.

**Comment by mdhaber** [comment]
> My intuition is to return the unconstrained MLE since there's no ambiguity.

I agree, and that is the problem with the current default for `boxcox_normmax`.

@lsorber @lorentzenchr Are you able to speak to the needs of scikit-learn or others w.r.t. these functions (`boxcox_normmax`/`yeojohnson_normmax` and related)? I've wanted to convert them to the array API, but the current interfaces have made it pretty difficult, and I'd like to talk it through in detail with someone who is interested.

**Comment by lorentzenchr** [comment]
`boxcox_normmax`/`yeojohnson_normmax` do not occur in scikit-learn.

**Comment by mdhaber** [comment]
@lorentzenchr My understanding from gh-10821 and gh-18389 was that there were features of sklearn that do these transforms. I was working on them with the impression that it would be nice to have one really solid version that is shared between the libraries. Do you mean that sklearn does not currently rely on SciPy for these transforms (i.e. it has its own implementation), or do you mean that sklearn only performs these transforms *given* $\lambda$ (but never performs MLE to find an optimal value of $\lambda$)?

**Comment by lsorber** [comment]
@mdhaber Quick response:
1. I believe sklearn now relies on scipy's implementation: https://github.com/scikit-learn/scikit-learn/issues/26308.
2. Observation: arguably, this PR replaces one magic constant (20) with another one (10000).
3. Observation:
    1. Before this PR, the implementation yields a transformation that does not overflow on _new_ data if that data is within a factor 20 of the training data.
    2. After this PR, there is no such guarantee anymore. The learned transformation may overflow on new data.
4. As a consequence of (3), I think this PR (as it is now) could cause serious problems for sklearn and other downstream packages. The use case for sklearn is learning the transformation, and then applying it to new data. And that will fail if no safety is built in to the transformation.

**Comment by mdhaber** [comment]
Thanks @lsorber. If we were to move forward with this PR, we would make sure it follows the precedent set by `boxcox_normmax` when it comes to overflow with new data. Are you sure this PR is not already maintaining that safety? See line 2015 and below where it emits a warning if "The optimal lambda is {lmax}, but the returned lambda is the "constrained optimum to ensure that the maximum or the minimum of the transformed data does not overflow...".

But this is the sort of thing I wanted to discuss - we added safety into these transformations, and I see that this is necessary for sklearn and other packages. But in retrospect, I don't think that sort of default is consistent with SciPy's usual philosophy. The function is documented to:

> Compute optimal Yeo-Johnson transform parameter for input data, using maximum likelihood estimation.

whereas what we're actually doing by default is:

> Compute a Yeo-Johnson transform parameter for input data using maximum likelihood estimation, subject to an undocumented constraint useful in machine learning applications, and printing a warning to the console with the true maximum likelihood estimate if the constraint is active.

IIRC, there was talk about adding an additional `xmax` parameter, which would complicate things further.

So as I'm considering how we can add `axis` and array API support to these functions, I wanted to discuss how/whether SciPy could meet all these needs. I don't really want to tackle this ATM, I just wanted to gather some information in the meantime.

- Would it be useful for these functions to support non-NumPy backends?
- Would it be useful for these functions to support vectorized calls over multiple slices of data?
- Were there other constraints (e.g. `xmax`) you wanted to be imposed on the optimization? Are there configuration options (e.g. magic number 20) that the user should be able to control?
- Is the `boxcox_normmax` `method='pearsonr'` useful to sklearn?

**Comment by xuefeng-xu** [comment]
> Would it be useful for these functions to support non-NumPy backends?
> Would it be useful for these functions to support vectorized calls over multiple slices of data?

I think it's useful.

> Is the boxcox_normmax method='pearsonr' useful to sklearn?

Currently `PowerTransformer` in scikit-learn uses `stats.boxcox`/`stats.yeojohnson`, which internally use `boxcox_normmax`/`yeojohnson_normmax`. And `stats.boxcox` uses `method='mle'` by default.

https://github.com/scikit-learn/scikit-learn/blob/d224a099ef535d74cc1d97d7795191448265d6ca/sklearn/preprocessing/_data.py#L3565

https://github.com/scikit-learn/scikit-learn/blob/d224a099ef535d74cc1d97d7795191448265d6ca/sklearn/preprocessing/_data.py#L3597

**Comment by lorentzenchr** [comment]
@xuefeng-xu Thanks for the clarification. My smartphone tricked me.

@mdhaber Array API support would be very nice for scikit-learn. I am not aware of other feature requests (in scikit-learn for yeo…/PowerTransformer).

**Comment by lsorber** [comment]
> Thanks @lsorber. If we were to move forward with this PR, we would make sure it follows the precedent set by `boxcox_normmax` when it comes to overflow with new data. Are you sure this PR is not already maintaining that safety? See line 2015 and below where it emits a warning if "The optimal lambda is {lmax}, but the returned lambda is the "constrained optimum to ensure that the maximum or the minimum of the transformed data does not overflow...".

I can categorically say that this PR does not maintain overflow safety on unseen data. And that's exactly the use case of sklearn. All the implementation guarantees is that the transformation does not overflow on the data provided to determine lambda itself.

> But this is the sort of thing I wanted to discuss - we added safety into these transformations, and I see that this is necessary for sklearn and other packages. But in retrospect, I don't think that sort of default is consistent with SciPy's usual philosophy. The function is documented to:
> 
> > Compute optimal Yeo-Johnson transform parameter for input data, using maximum likelihood estimation.
> 
> whereas what we're actually doing by default is:
> 
> > Compute a Yeo-Johnson transform parameter for input data using maximum likelihood estimation, subject to an undocumented constraint useful in machine learning applications, and printing a warning to the console with the true maximum likelihood estimate if the constraint is active.

I'm a bit conflicted on this:
1. On the one hand, I fully understand the viewpoint that the scipy implementation should be straightforward and aligned with the transformation as it is mathematically defined.
2. On the other hand, any practical use of the transformation will want to constrain lambda so that the transformation does not overflow on the given data (from which lambda is learned), nor on unseen data (on which the learned transformation is applied).

Unfortunately, the risk of overflow (which is substantial, as simple examples demonstrate) in my opinion requires scipy to make a choice: either it implements the transformation as it is mathematically defined and leaves it up to the downstream packages and users to be aware of this issue and manage it, or it offers options to safeguard against overflow.

My personal preference is with the latter option, whereby the constraint on lambda is only just what is needed to prevent overflow.

This whole situation arises because the transformation itself is ill-posed.  The problem disappears entirely if you regularise lambda (cf. https://github.com/scipy/scipy/issues/18389#issuecomment-1537089805, 'item 2'). This could be a third option, with the downside that such regularisation is non-standard.

> IIRC, there was talk about adding an additional `xmax` parameter, which would complicate things further.

As discussed above, I personally think that would be quite reasonable.

> So as I'm considering how we can add `axis` and array API support to these functions, I wanted to discuss how/whether SciPy could meet all these needs. I don't really want to tackle this ATM, I just wanted to gather some information in the meantime.
> 
> * Would it be useful for these functions to support non-NumPy backends?
> * Would it be useful for these functions to support vectorized calls over multiple slices of data?
> * Were there other constraints (e.g. `xmax`) you wanted to be imposed on the optimization? Are there configuration options (e.g. magic number 20) that the user should be able to control?
> * Is the `boxcox_normmax` `method='pearsonr'` useful to sklearn?

Not sure if you're asking my opinion on this, but FWIW:
- I would think so, yes.
- I would think so, yes.
- I think `xmax` would be the most important one, with the downside that it would have to be a vector as well if we go with this solution. A safety factor could be a single argument for all vectorised applications of the transformation.
- I would think so, yes.

I would add that I expect that none of these features are important relative to other improvements scipy may need. It is my impression that power transforms are not all that widely used by sklearn users.

**Comment by mdhaber** [comment]
> I can categorically say that this PR does not maintain overflow safety on unseen data.

Ok, I hadn't checked. I agree that we wouldn't want to go backwards on that accidentally.

> It is my impression that power transforms are not all that widely used by sklearn users

Hmm... These features have been the subject of a fair number of pretty technical PRs over the course of several years, with many more needed/wanted. I have supported them because sometimes it is interesting, but then I wonder whether it is the best use of the time, if nobody is using the features and there are a few dozen other PRs in the works. 

Thanks for the feedback, all. Let me mull over this for a few days.

**Comment by lsorber** [comment]
> > It is my impression that power transforms are not all that widely used by sklearn users
> 
> Hmm... These features have been the subject of a fair number of pretty technical PRs over the course of several years, with many more needed/wanted. I have supported them because sometimes it is interesting, but then I wonder whether it is the best use of the time, if nobody is using the features and there are a few dozen other PRs in the works.

Just to clarify my statement and thoughts on this:
- I think that in practice, many sklearn pipelines don't make use of a power transform. They don't add any value to decision tree-based models for instance.
- That said, I'm also quite sure the power transform is still used in a nontrivial fraction of sklearn pipelines. It's not unimportant—it's just not a standard tool that data scientists would reach for.
- I think the right thing to do is to improve it to the point where it's easy and safe to use the transform so that people can use it without fear of it breaking their pipelines. That will improve adoption, and possibly drive the need for further features in the future.
- I think all of the work that has been put into it so far has been valuable, and not at all a waste of effort. It was necessary and good in my opinon.

Unrelated to this issue, I can imagine it's not always easy to navigate these decisions and priorities, and I'd like to say that I appreciate everything you do for this project. Scipy is lucky to have you and the care that you put into it!

**Comment by mdhaber** [comment]
Thanks @lsorber!

---

I took a careful look at what the different Box-Cox functions currently do. (The story with Yeo-Johnson is similar, but fortunately a little simpler.) Roughly:

`boxcox_llf`:
- Evaluates the Box-Cox log-likelihood for a single value of $\lambda$

`boxcox_normplot`:
- Evaluates the Pearson correlation coefficient for multiple values of $\lambda$
- Plots the Pearson correlation coefficient vs $\lambda$

`boxcox_normmax`:
- Finds $\lambda$ that maximizes the Box-Cox log-likelihood or
- Finds $\lambda$ that maximizes the Pearson correlation coefficient or
- Finds and returns both (separately)

`boxcox`:
- If $\lambda$ is provided, calls `scipy.special.boxcox`
- If $\lambda$ is not provided, calls `boxcox_normmax` and `scipy.special.boxcox`
- If $\lambda$ is not provided and $\alpha$ is provided,  calls `boxcox_normmax`, `scipy.special.boxcox`, and a private function for computing a confidence interval

Some issues are:
- The number of outputs of functions change depending on the input parameters: `boxcox_normmax` can return one or two values of $\lambda$ depending on `method`, and `boxcox` may return one, two, or three different values, the last of which is a actually a two-tuple. Among other things, this makes it challenging to apply the `_axis_nan_policy` decorator and run related property-based tests.
- The functions miss the opportunity to be modular in a few ways. For instance:
  - When $\lambda$ is not provided, `boxcox` calls `boxcox_normmax` with `method='mle'` (only), and when `alpha` is provided, it calls a private function to compute the confidence interval. I'd suggest that computing the optimal $\lambda$ (according to MLE, Pearson correlation, or some other objective) and corresponding confidence interval should be kept separate from the comparatively simpler task of transforming the data given $\lambda$, which is already covered by `scipy.special.boxcox`. 
  - `boxcox_normplot` allows computing the Pearson correlation coefficient for multiple values of $\lambda$ and combines that with plotting, whereas `boxcox_llf` computes the LLF for only a single value of $\lambda$. ISTM that we would like to be able to compute the LLF for multiple values of $\lambda$ efficiently, and we might want to produce plots of either relationship.
  - `boxcox_normmax` and `boxcox` accept an optimizer, but it would be about as simple for the user - and more flexible for the sorts of customizations sklearn needs, and much easier to extend these functions in the future - if we just provided the primitives needed for the user to call their optimizer of choice.

I think users and dependent libraries would be better served if SciPy provided the following functionality:
- A function that computes the Box-Cox transform *given* $\lambda$. This is already provided by [`scipy.special.boxcox`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.boxcox.html#scipy.special.boxcox).
- A function that computes the inverse of the transform w.r.t. `y`. (It looks like sklearn has its own implementation of this?) This is already provided by [`scipy.special.inv_boxcox`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.inv_boxcox.html#scipy.special.inv_boxcox).
- A function that computes the inverse of the transform w.r.t. $\lambda$. This can be used to find bounds of $\lambda$ for use in optimization. We have a private function for this, but it would be nice to make it a public ufunc like `scipy.special.boxcox` and `scipy.special.inv_boxcox`.
- Functions that compute the common objectives used to optimize $\lambda$ - e.g. log-likelihood and Pearson correlation coefficient. Ideally, these would allow efficient, vectorized calls w.r.t. $\lambda$ for faster optimization and plotting. The obvious thing to do would be to vectorize `boxcox_llf` and add a similar function for the Pearson correlation coefficient.
- (Potentially) a convenience function for plotting the effect of $\lambda$ on the objective function(s). Personally, I'd rather leave this to user-land; the vectorized objective functions would avoid the need to do cumbersome looping, and all that is really left is adding axis labels and a title.
- (Potentially) a function that performs the optimization - although I would prefer to just provide a recipe so users and dependent libraries can perform the optimization with their desired constraints in just a few lines. This avoids the trouble we've had with magic numbers and an ever-expanding interface that *still* doesn't meet all needs.
- (Potentially) a function that computes a confidence interval for $\lambda$ - but I need to read up on the underlying assumptions and how this is computed. 

Note that in user-land, performing bounded MLE with custom `xmax` and `ymax` would be a two-step process.

1. Use the special function that inverts the transform w.r.t. $\lambda$ to compute bounds on $\lambda$. ISTM there doesn't need to be a two-step process like we have in the current code: the smallest `x` value and most negative permissible `y` value gives the lower bound of $\lambda$, and the maximum `x` and `y` values give the upper bound of $\lambda$.
2. Use one of SciPy's bounded scalar minimization routines with these bounds and the desired objective function.

To use a regularized objective function, only the first step changes - use `optimize.elementwise.bracket_minimum` to find a valid bracket for minimization.

I'd suggest a similar set of functions for Yeo-Johnson. If we allow ourselves to start fresh like this, it would be **much** easier to make everything vectorized and array API compatible without disturbing backward compatibility of the existing functions, and we'd end up with a simpler yet more powerful interface.

What do you think?

**Comment by xuefeng-xu** [comment]
+1

> (It looks like sklearn has its own implementation of this?) This is already provided by [scipy.special.inv_boxcox](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.inv_boxcox.html#scipy.special.inv_boxcox).

scikit-learn already changes to use `scipy.special.inv_boxcox` since their own implementation has a bug.

https://github.com/scikit-learn/scikit-learn/blob/a885b523630f0e2430c214ea30a1b3a74750dcaa/sklearn/preprocessing/_data.py#L3489

But for yeojohnson and its inverse, scikit-learn still uses its own implementation.

https://github.com/scikit-learn/scikit-learn/blob/a885b523630f0e2430c214ea30a1b3a74750dcaa/sklearn/preprocessing/_data.py#L3440

https://github.com/scikit-learn/scikit-learn/blob/a885b523630f0e2430c214ea30a1b3a74750dcaa/sklearn/preprocessing/_data.py#L3490

scipy currently doesn't have `inv_yeojohnson`, maybe we can add one.
